dataforest/core/Spec.py

```diff
@@ -187,41 +189,22 @@ def _build_run_spec_lookup(self) -> Dict[str, "RunSpec"]:
             run_spec_lookup[process_name] = run_spec
         return run_spec_lookup
 
-    def _build_precursors_lookup(self) -> Dict[str, List[str]]:
+    def _build_precursors_lookup(self, incl_root: bool = False, incl_current: bool = False) -> Dict[str, List[str]]:
         """See class definition"""
-        precursors = {"root": []}
         current_precursors = []
+        if incl_root and incl_current:
+            current_precursors = current_precursors + ["root"]
+        precursors = {"root": current_precursors}
+        if incl_root and not incl_current:
+            current_precursors = current_precursors + ["root"]
```

Comment:
I don't think this will affect too much for small lists, but it terms of time complexity, this concatenation create a new list object each time (O(n)), whereas `.append()` will just put the last element in a contiguous pre-allocated array memory slot (O(1)).
```
current_precursors.append("root")
```

Same in line 196

Reply:
Glad to see that you're looking out for time complexity! In this case, I think you're right that it's negligible due to the small size of the lists. I originally used `append`, but then I had to write an extra `copy` line so that I wasn't modifying the list I'd already put into dict, so I did this for readability.

Reply:
Gotcha!